### PR TITLE
refactor: #163 - Introduce controller runtime for watching top-level CRDs

### PR DESCRIPTION
**Pull request by fpetkovski**
Metacontroller, being an older project, contains a lot of boilerplate
for watching and fetching kubernetes resources. Most of that logic
has been extracted into sigs.k8s.io/controller-runtime.

This commit introduces controller-runtime for reconciling top level
resources, that is DecoratorControllers and CompositeControllers.
Metacontroller creates one controller for each of these resources when
it starts. The scope of this commit is refactoring these particular controllers
to take advantage of the machinery provided by controller-runtime.

In addition to the two static controllers, metacontroller creates a new dynamic
controller for each instance of DecoratorController and CompositeController.
These dynamic controllers have been left untouched and their refactoring
to controller-runtime, or to any other library for that manner, is out of scope
for this particular commit. That work would be done in a subsequent commit.

Since controller-runtime uses a dynamic client, these cnages make the
informes and listers for CompositeController and DecoratorController
under client/generated obsolete. Their removal can also be done as follow up
once this commit has been merged.

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

### CODE REVIEW

File: main.go
Comment (grzesuav) [inline main.go:65]: should be `ErrorS` for structured logging

**Comment by grzesuav** [comment]
Thanks @fpetkovski ! Will check out the code and take a look in the week (will try before Easter), probably will have some question related to how controller runtime works itself.

In general if some parts of code are not needed I would prefer to remove them as part of the refactoring as later it can be lost. It would also allow to make deployments a bit faster as we wouldn't need to generate code.

**Comment by grzesuav** [comment]
Also, does controller  runtime allows to implement sth likehttps://github.com/metacontroller/metacontroller/issues/218 ? I mean to specify label selectors for managed resources ?

**Comment by grzesuav** [comment]
Also, for  fixing imports and formatting , you can run those two commands :
* `gofmt -w .  `
* `goimports  -w -l . `

**Comment by grzesuav** [comment]
also, if you could hold on for few days with adjusting this PR, we have two other ones in review and they impact the same code area, do not want to make you tired with contently adjusting this PR, so best wait till two other ones (emit 
 kubernetes events and runtimeContext) will be merged

**Comment by fpetkovski** [comment]
Hey @grzesuav, thanks for the fast turnaround.

> In general if some parts of code are not needed I would prefer to remove them as part of the refactoring as later it can be lost. It would also allow to make deployments a bit faster as we wouldn't need to generate code.
Sounds good, will remove the extra files.

> Also, does controller runtime allows to implement sth like #218 ? I mean to specify label selectors for managed resources ?
I am not entirely sure, but I can take a look if we can set label selectors on watches.


> also, if you could hold on for few days with adjusting this PR, we have two other ones in review and they impact the same code area, do not want to make you tired with contently adjusting this PR, so best wait till two other ones (emit
> kubernetes events and runtimeContext) will be merged
No worries, will wait until those ones are merged before proceeding.

File: controller/composite/metacontroller.go
Comment (AmitKumarDas) [inline controller/composite/metacontroller.go:39]: IMO this commit should be complimented with missing code level comments or explanations.
It helps the community to understand better.

File: controller/composite/metacontroller.go
Comment (AmitKumarDas) [inline controller/composite/metacontroller.go:75]: Do you think we should be more verbose in the naming conventions.
I was thinking for a moment whose name are we talking about.

File: controller/composite/metacontroller.go
Comment (AmitKumarDas) [inline controller/composite/metacontroller.go:79]: Should we write in an idiomatic style. For example:
```go
cc = v1alpha1.CompositeController{}
err := mc.k8s.Get(ctx.Background(), request.NamespacedName, &cc)
```

File: controller/composite/metacontroller.go
Comment (AmitKumarDas) [inline controller/composite/metacontroller.go:74]: Does this need to be named as `Reconcile`?
Can we have this as `Sync` & rename `syncCompositeController` to `sync`
Or we continue having this as `Reconcile` & rename `syncCompositeController` to `reconcile`
It may be more appropriate to think these as separate PRs as well.
I was thinking more in terms of readability, idiomatic conventions wherever possible.

**Comment by grzesuav** [comment]
> > Also, does controller runtime allows to implement sth like #218 ? I mean to specify label selectors for managed resources ?
> 
> I am not entirely sure, but I can take a look if we can set label selectors on watches.
> 

Just checked, there are predicates on `Watch` which can be used

File: controller/composite/metacontroller.go
Comment (grzesuav) [reply controller/composite/metacontroller.go:74]: I am voting to renaming `sync ` to `reconcileCompositeController` as it seems it is the term used among controllers

Comment (fpetkovski) [reply controller/composite/metacontroller.go:85]: The name needs to be `Reconcile` since it is the one defined by the interface. However I renamed the non-exported `sync` methods to `reconcileCompositeController` and `reconcileDecoratorController` respectively.

File: main.go
Comment (fpetkovski) [inline main.go:48]: The `controller-runtime` package will look for an optional `--kubeconfig` so this one should not be needed anymore. Changing the flag name is a breaking change and we can do one of two things
- Keep the flag and somehow adapt controller-runtime to use this one
- Document the breaking change in the release notes

File: server/server.go
Comment (fpetkovski) [inline server/server.go:36]: This interface is not needed anymore because `controller-runtime` already defines a `Reconciler` interface which is implemented by all controllers. Stopping the controllers managed by the manager is done automatically when the manager itself is stopped.

File: server/server.go
Comment (fpetkovski) [inline server/server.go:66]: It is usually a good practice to decouple initiating and starting managers. So I decided to move the logic to start and terminate the manager into main.go

File: test/integration/go.mod
Comment (fpetkovski) [inline test/integration/go.mod:13]: The reason for using v0.5.4 is because it requires the same kubernetes library versions as the ones defined in metacontroller's `go.mod`. As soon as we update our own kubernetes library versions, we can upgrade controller-runtime as well

Comment (grzesuav) [reply test/integration/go.mod:13]: oh, this raises a important question, does controller-runtume limits somehow on which k8s versions it can run ? As currently metacontrollers support vast range of k8s versions, by not depending on k8s API directly.

File: main.go
Comment (grzesuav) [reply main.go:48]: if its possible I would took non-breaking change approach, like marking flag as deprecated for now and pass it to controller-runtime, but would need to see whole PR, so please hold with adjusting that for now

File: controller/composite/metacontroller.go
Comment (grzesuav) [reply controller/composite/metacontroller.go]: I just wonder if we need another client or we can reuse `dynClient` already present there- i.e. :
```go
	resourceClient, _ := mc.dynClient.Kind(cc.APIVersion, cc.Kind)
	_, err := resourceClient.Namespace(request.Namespace).Get(request.Name, metav1.GetOptions{})
```
I will try it out locally and let you know

File: controller/composite/metacontroller.go
Comment (grzesuav) [reply controller/composite/metacontroller.go:85]: Actually I think changing the name can be good, as `reconcile` is widely used term of syncing resources in operator world

File: controller/composite/metacontroller.go
Comment (grzesuav) [reply controller/composite/metacontroller.go]: Was able to do it via 
```go
resourceClient, err := mc.dynClient.Kind("metacontroller.k8s.io/v1alpha1", "CompositeController")
```
but it returns `Unstructured`. I will experiment a bit as  it it not very suitable I guess

Comment (fpetkovski) [reply controller/composite/metacontroller.go]: I think another difference is that the dynamic client is queuing the apiserver directly, while the one from controller-runtime is querying a cache. The cache is updates automatically in the background for each object that the controller is watching.

File: test/integration/go.mod
Comment (fpetkovski) [reply test/integration/go.mod:13]: This is a good question. I am not aware of any limitation controller-runtime itself has outside of the ones that come from the kubernetes libraries. Seeing green e2e tests is also a good indication that things would continue working with older versions. 

> As currently metacontrollers support vast range of k8s versions, by not depending on k8s API directly.

Could you elaborate a bit on this point? How does metacontroller avoid a direct dependency on the k8s api?

Comment (grzesuav) [reply test/integration/go.mod:13]: currently we depend on k8s core libraries but nothing related to typing, except `Events`, but their api was stable from early version of k8s

Comment (grzesuav) [reply test/integration/go.mod:13]: all of the code usues `Unstructured` when recieving and sending types to API server

File: server/server.go
Comment (grzesuav) [inline server/server.go:43]: @fpetkovski can we :
1. move `mcClient` declaration before
2. move `controllerContext` declaration before
3. refactor `ControllerContext` in common package, to have `broadcaster` public
4. Set `EventBroadcaster` instance in `manager.Options` to the same instance
5. (Optional) if it would be  possible, replace and remove the one created in the `ControllerContext`

main motivation is that we have options to limit `spam level` of broadcaster in input flags and would be good to have all instances using the same limit

Comment (grzesuav) [reply server/server.go:43]: if removal of mode from `ControllerContext` will be to big/to complicated leave it, it can be removed later

File: test/integration/go.mod
Comment (grzesuav) [reply test/integration/go.mod:13]: but I think that for metacontroller resources we are safe - those CRD's are defined in the controller itself. 

As for the rest - maybe there is no point keeping resources in older version - I decoupled test dependencies to test sub-project, I will test if we are really independent

**Review by grzesuav — CHANGES_REQUESTED** [review]
As mentioned in comment would appreciate if :
1. we can take non-breaking approach for client config flag
2. adjust `EventBroadcaster` to be the same instance as in `ControllerContext

**Comment by grzesuav** [comment]
btw @fpetkovski nice testing framework improvements ! - should be more stable now

File: server/server.go
Comment (fpetkovski) [reply server/server.go]: Alright, 1-4 should be done. 

I did not quite understand point 5. What specifically would you suggest we replace and remove in `ControllerContext`?

**Review by grzesuav — APPROVED** [review]
Looks good from my side, but want @AmitKumarDas to look for some go-practices

**Comment by grzesuav** [comment]
As discussed with Amit, if sth need to be adjusted I will do it in post-MR

**Pull request merged**
